tests/DCPS/GroupPresentation/SubscriberListener.cpp

```diff
@@ -127,66 +127,52 @@ SubscriberListenerImpl::on_data_on_readers(
         Messenger::MessageDataReader::_narrow(readers[i]);
 
       Messenger::MessageSeq msg;
       ::DDS::SampleInfoSeq si;
-      ret = message_dr->take(msg, si,
-              DDS::LENGTH_UNLIMITED,
-              ::DDS::NOT_READ_SAMPLE_STATE,
-              ::DDS::ANY_VIEW_STATE,
-              ::DDS::ANY_INSTANCE_STATE) ;
-
-      if (si[0].instance_state == DDS::NOT_ALIVE_DISPOSED_INSTANCE_STATE
-         || si[0].instance_state == DDS::NOT_ALIVE_NO_WRITERS_INSTANCE_STATE)
-
-      if (msg.length() != num_messages || si.length() != num_messages) {
+      ret = message_dr->take(msg, si, DDS::LENGTH_UNLIMITED,
+                             ::DDS::NOT_READ_SAMPLE_STATE,
+                             ::DDS::ANY_VIEW_STATE,
+                             ::DDS::ANY_INSTANCE_STATE);
+      if (ret != DDS::RETCODE_OK) {
+        if (ret == DDS::RETCODE_NO_DATA) {
+          ACE_ERROR((LM_ERROR,ACE_TEXT("%N:%l:%t: data reader[%d] has no data\n"),readers[i]->get_instance_handle()));
+          continue;
+        }
         ACE_ERROR((LM_ERROR,
-                    ACE_TEXT("%N:%l: SubscriberListenerImpl::on_data_on_readers()")
-                    ACE_TEXT(" ERROR: MessageSeq %d SampleInfoSeq %d !\n"),
-                    msg.length(), si.length()));
-        this->verify_result_ = false;
+                   ACE_TEXT("%N:%l:%t: SubscriberListenerImpl::on_data_on_readers()")
+          ACE_TEXT(" ERROR: reader[%d] failed with %C\n"),
+                     readers[i]->get_instance_handle(), OpenDDS::DCPS::retcode_to_string(ret)));
+          ACE_OS::exit(-1);
       }
 
-      if (ret != DDS::RETCODE_OK) {
+      CORBA::ULong num_samples = si.length();
+      if (num_samples > 1 && (msg.length() != num_samples || num_samples < num_messages * 2)) {
         ACE_ERROR((LM_ERROR,
-                    ACE_TEXT("%N:%l: SubscriberListenerImpl::on_data_on_readers()")
-                    ACE_TEXT(" ERROR: read failed!\n")));
-        ACE_OS::exit(-1);
+                   ACE_TEXT("%N:%l: SubscriberListenerImpl::on_data_on_readers()")
+        ACE_TEXT(" ERROR: MessageSeq %d SampleInfoSeq %d != %d\n"),
+                   msg.length(), si.length(), num_messages));
+        this->verify_result_ = false;
       }
 
-      CORBA::ULong num_samples = si.length();
       for (CORBA::ULong i = 0; i < num_samples; ++i) {
-        this->verify (msg[i], si[i], qos, i == num_samples - 1 ? true : false);
+        this->verify (msg[i], si[i], qos, i == (num_samples - 1));
       }
     }
   }
   else { //::DDS::INSTANCE_PRESENTATION_QOS
     subs->notify_datareaders ();
   }
-
   ret = subs->end_access ();
   if (ret != ::DDS::RETCODE_OK) {
     ACE_ERROR((LM_ERROR,
-                  ACE_TEXT("%N:%l: SubscriberListenerImpl::on_data_on_readers()")
-                  ACE_TEXT(" ERROR: end_access failed!\n")));
+               ACE_TEXT("%N:%l: SubscriberListenerImpl::on_data_on_readers()")
+    ACE_TEXT(" ERROR: end_access failed!\n")));
```

Comment:
I think this line should line up with the one above, since it's still part of the ACE_ERROR message.

Reply:
Yeah, that's my editor being "helpful."